# How to use csvkit
### What is csvkit?
csvkit is a suite of command-line tools for converting to and working with CSV.
### Why csvkit?
For many different types of data processing operations, all we need is a single line command using csvkit.
## 1. Getting started
### 1.1. Installing csvkit
csvkit can be easily installed using pip:  

>$ pip install csvkit

### 1.2. Fetching the data
This tutorial will introduce how to use csvkit by analyzing real datasets: One dataset [Fish Springs National Wildlife Refuge Odonata records 2015](https://catalog.data.gov/dataset/fish-springs-national-wildlife-refuge-odonata-records-2015) is in Excel format, another dataset [Most Popular Baby Girl Names, 1980-2013](https://catalog.data.gov/dataset/most-popular-baby-girl-names-1980-2013-3f00a) is in CSV format. Use curl to fetch and rename them to odonata.xlsx and name.csv:

In [1]:
!curl -o odonata.xlsx https://ecos.fws.gov/ServCat/DownloadFile/102067?Reference=61548
!curl -o name.csv https://data.illinois.gov/api/views/rx64-7qfr/rows.csv?accessType=DOWNLOAD

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 11905  100 11905    0     0   8208      0  0:00:01  0:00:01 --:--:--  8564
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 16322    0 16322    0     0  49328      0 --:--:-- --:--:-- --:--:-- 74871


### 1.3. A core feature: pipeability
All the csvkit tools take input from standard input and write output to standard output, which means the output of one csvkit tool can become the input of another. Combined with the standard command-line pipe, csvkit tools can form a data processing pipeline without creating unnecessary intermediary files. Keep this feature in mind and we will demonstrate the power of "csvkit plus pipe" later.

## 2. Basic usage

csvkit tools can be categorized into three main types: input, output, and processing.

### 2.1. Input
### in2csv
in2csv converts various tabular data formats (csv, fixed, geojson, json, ndjson, xls, xlsx) into CSV format.

It's convenient to get usage information about a csvkit tool with "-h" argument:

In [2]:
!in2csv -h

usage: in2csv [-h] [-d DELIMITER] [-t] [-q QUOTECHAR] [-u {0,1,2,3}] [-b]
              [-p ESCAPECHAR] [-z MAXFIELDSIZE] [-e ENCODING] [-S] [-H] [-v]
              [-l] [--zero] [-f FILETYPE] [-s SCHEMA] [-k KEY] [-y SNIFFLIMIT]
              [--sheet SHEET] [--no-inference]
              [FILE]

Convert common, but less awesome, tabular data formats to CSV.

positional arguments:
  FILE                  The CSV file to operate on. If omitted, will accept
                        input on STDIN.

optional arguments:
  -h, --help            show this help message and exit
  -d DELIMITER, --delimiter DELIMITER
                        Delimiting character of the input CSV file.
  -t, --tabs            Specifies that the input CSV file is delimited with
                        tabs. Overrides "-d".
  -q QUOTECHAR, --quotechar QUOTECHAR
                        Character used to quote strings in the input CSV file.
  -u {0,1,2,3}, --quoting {0,1,2,3}
                        Quoting style use

We can convert odonata.xlsx from Excel into CSV format:

In [3]:
!in2csv odonata.xlsx

Common Name,Family,Scientific Name,State,County,Location,Lat.,Long.,Elev.,Date,Collector,#,Repository,Determination
Common Green Darner,Aeshnidae,Anax junius,Utah,Juab,Baker Hot Springs (7 miles north of Brush Wellman Road),39.61039,-112.73041,4628 ft,2015-08-13,"Bybee, Clark & Myrup",2,BYUC (DNA Tissue),Myrup 2015
Paiute Dancer,Coenagrionidae,Argia alberta,Utah,Juab,Baker Hot Springs (7 miles north of Brush Wellman Road),39.61039,-112.73041,4628 ft,2015-08-13,"Bybee, Clark & Myrup",11,BYUC (DNA Tissue),Myrup 2015
Black-fronted Forktail,Coenagrionidae,Ischnura denticollis,Utah,Juab,Baker Hot Springs (7 miles north of Brush Wellman Road),39.61039,-112.73041,4628 ft,2015-08-13,"Bybee, Clark & Myrup",15,BYUC (DNA Tissue),Myrup 2015
Western Pondhawk,Libellulidae,Erythemis collocata,Utah,Juab,Baker Hot Springs (7 miles north of Brush Wellman Road),39.61039,-112.73041,4628 ft,2015-08-13,"Bybee, Clark & Myrup",3,BYUC (DNA Tissue),Myrup 2015
Comanche Skimmer,Libellulidae,Libellula comanche,Uta

Instead of sending the output to the terminal, it's more useful to write the output to a CSV file:

In [4]:
!in2csv odonata.xlsx > odonata.csv

If we run in2csv for a CSV file, the formatting (quoting, line endings, etc.) of the file will be standardized. However, it will be too long to display all the output. Remember we can combine pipe with csvkit? Let's use the head command to display the first sixty lines:

In [5]:
!in2csv name.csv | head -n60

Rank,Year,Name,Frequency
1,1980,Jennifer,3017
2,1980,Amanda,1750
3,1980,Melissa,1566
4,1980,Sarah,1390
5,1980,Jessica,1373
6,1980,Nicole,1336
7,1980,Elizabeth,1221
8,1980,Michelle,1170
9,1980,Amy,1013
10,1980,Tiffany,941
11,1980,Angela,924
12,1980,Kelly,895
13,1980,Heather,893
14,1980,Kimberly,854
15,1980,Lisa,845
16,1980,Stephanie,839
17,1980,Christina,778
18,1980,Rebecca,757
19,1980,Laura,756
20,1980,Erin,736
21,1980,Mary,655
22,1980,Jamie,630
23,1980,Megan,604
24,1980,Rachel,603
25,1980,Sara,601
1,1981,Jennifer,2920
2,1981,Jessica,1683
3,1981,Amanda,1607
4,1981,Sarah,1565
5,1981,Melissa,1401
6,1981,Nicole,1341
7,1981,Elizabeth,1221
8,1981,Michelle,1024
9,1981,Amy,1015
10,1981,Stephanie,976
11,1981,Tiffany,921
12,1981,Rebecca,893
13,1981,Kimberly,805
14,1981,Angel,804
15,1981,Lisa,778
16,1981,Erin,771
17,1981,Heather,770
18,1981,Laura,742
19,1981,Kelly,708
20,1981,Christina,636
21,1981,Sara,625
22,1981,Mary,624
23,1981,Kristin,599
24,1981,Rachel,596
25,1981,Jamie,590
1,1982,Jennifer,

### 2.2. Output
### csvlook
csvlook renders a CSV file to the command line in a readable format.  

Let's take a look at name.csv:

In [6]:
!csvlook name.csv | head

|-------+------+-----------+------------|
|  Rank | Year | Name      | Frequency  |
|-------+------+-----------+------------|
|  1    | 1980 | Jennifer  | 3017       |
|  2    | 1980 | Amanda    | 1750       |
|  3    | 1980 | Melissa   | 1566       |
|  4    | 1980 | Sarah     | 1390       |
|  5    | 1980 | Jessica   | 1373       |
|  6    | 1980 | Nicole    | 1336       |
|  7    | 1980 | Elizabeth | 1221       |


However, if a table is too wide, the output of csvlook may not be easy to read:

In [7]:
!csvlook odonata.csv | head -n6

|-------------------------+----------------+--------------------------+-------+--------+---------------------------------------------------------+----------+------------+---------+------------+----------------------+----+-------------------+----------------|
|  Common Name            | Family         | Scientific Name          | State | County | Location                                                | Lat.     | Long.      | Elev.   | Date       | Collector            | #  | Repository        | Determination  |
|-------------------------+----------------+--------------------------+-------+--------+---------------------------------------------------------+----------+------------+---------+------------+----------------------+----+-------------------+----------------|
|  Common Green Darner    | Aeshnidae      | Anax junius              | Utah  | Juab   | Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft | 2015-08-13 | Bybee, Clark & Myrup | 2  | B

A tool to truncate the table would be helpful which will be introduced later.

### csvformat
csvformat converts a CSV file to a custom output format.  

We can convert name.csv from a comma-delimited to a pipe-delimited one:

In [8]:
!csvformat -D "|" name.csv | head

Rank|Year|Name|Frequency
1|1980|Jennifer|3017
2|1980|Amanda|1750
3|1980|Melissa|1566
4|1980|Sarah|1390
5|1980|Jessica|1373
6|1980|Nicole|1336
7|1980|Elizabeth|1221
8|1980|Michelle|1170
9|1980|Amy|1013


Try to take the output of csvformat as the input of in2csv to see the result of standardizing the formatting of the CSV file: 

In [9]:
!csvformat -D "|" name.csv | head | in2csv

usage: in2csv [-h] [-d DELIMITER] [-t] [-q QUOTECHAR] [-u {0,1,2,3}] [-b]
              [-p ESCAPECHAR] [-z MAXFIELDSIZE] [-e ENCODING] [-S] [-H] [-v]
              [-l] [--zero] [-f FILETYPE] [-s SCHEMA] [-k KEY] [-y SNIFFLIMIT]
              [--sheet SHEET] [--no-inference]
              [FILE]
in2csv: error: You must specify a format when providing data via STDIN (pipe).


Oops... What's wrong? The error message told us how to fix it. For in2csv, we have to specify a format when providing data via standard input:

In [10]:
!csvformat -D "|" name.csv | head | in2csv -f csv

Rank,Year,Name,Frequency
1,1980,Jennifer,3017
2,1980,Amanda,1750
3,1980,Melissa,1566
4,1980,Sarah,1390
5,1980,Jessica,1373
6,1980,Nicole,1336
7,1980,Elizabeth,1221
8,1980,Michelle,1170
9,1980,Amy,1013


### csvstat
csvstat intelligently determines the type of each column and performs basic statistics for all columns.

Let's examine the summary statistics for name.csv:

In [11]:
!csvstat name.csv

  1. Rank
	<type 'int'>
	Nulls: False
	Min: 1
	Max: 25
	Sum: 11075
	Mean: 13.0141010576
	Median: 13
	Standard Deviation: 7.21858083478
	Unique values: 25
	5 most frequent values:
		25:	35
		24:	34
		20:	34
		21:	34
		22:	34
  2. Year
	<type 'int'>
	Nulls: False
	Min: 1980
	Max: 2013
	Sum: 1699011
	Mean: 1996.48766157
	Median: 1996
	Standard Deviation: 9.81153907382
	Unique values: 34
	5 most frequent values:
		1986:	26
		1987:	25
		1984:	25
		1985:	25
		1982:	25
  3. Name
	<type 'unicode'>
	Nulls: False
	Unique values: 102
	5 most frequent values:
		Sarah:	26
		Jessica:	26
		Elizabeth:	26
		Jennifer:	24
		Lauren:	24
	Max length: 9
  4. Frequency
	<type 'int'>
	Nulls: False
	Min: 264
	Max: 3017
	Sum: 646676
	Mean: 759.901292597
	Median: 664
	Standard Deviation: 387.063122755
	Unique values: 586
	5 most frequent values:
		592:	5
		630:	5
		600:	5
		594:	4
		596:	4

Row count: 851


We can specify the columns we care about and only get the statistics of these columns:

In [12]:
!csvstat -c 1,2 name.csv

  1. Rank
	<type 'int'>
	Nulls: False
	Min: 1
	Max: 25
	Sum: 11075
	Mean: 13.0141010576
	Median: 13
	Standard Deviation: 7.21858083478
	Unique values: 25
	5 most frequent values:
		25:	35
		24:	34
		20:	34
		21:	34
		22:	34
  2. Year
	<type 'int'>
	Nulls: False
	Min: 1980
	Max: 2013
	Sum: 1699011
	Mean: 1996.48766157
	Median: 1996
	Standard Deviation: 9.81153907382
	Unique values: 34
	5 most frequent values:
		1986:	26
		1987:	25
		1984:	25
		1985:	25
		1982:	25

Row count: 851


We can also replace the column numbers with column names and get the exactly same result:

In [13]:
!csvstat -c Rank,Year name.csv

  1. Rank
	<type 'int'>
	Nulls: False
	Min: 1
	Max: 25
	Sum: 11075
	Mean: 13.0141010576
	Median: 13
	Standard Deviation: 7.21858083478
	Unique values: 25
	5 most frequent values:
		25:	35
		24:	34
		20:	34
		21:	34
		22:	34
  2. Year
	<type 'int'>
	Nulls: False
	Min: 1980
	Max: 2013
	Sum: 1699011
	Mean: 1996.48766157
	Median: 1996
	Standard Deviation: 9.81153907382
	Unique values: 34
	5 most frequent values:
		1986:	26
		1987:	25
		1984:	25
		1985:	25
		1982:	25

Row count: 851


If we only need one certain kind of statistics, we can specify the statistic argument. For example, we can get the number of unique values of each column by passing the statistic name "unique":

In [14]:
!csvstat --unique name.csv

  1. Rank: 25
  2. Year: 34
  3. Name: 102
  4. Frequency: 586


From the above result, we can learn that the data in name.csv contains the top 25 popular baby girl names in each of the 34 years. Since there are only 102 different names, we can also learn that many popular names must repeatedly appear in several years.

Maybe the unique statistics for frequency doesn't make much sense for now, so let's combine the two filters together to just get the unique information about the first three columns:

In [15]:
!csvstat -c 1,2,3 --unique name.csv

  1. Rank: 25
  2. Year: 34
  3. Name: 102


### 2.3. Processing
### csvclean
csvclean cleans a CSV file of common syntax errors.  

Why not use csvclean to test name.csv and odonata.csv:

In [16]:
!csvclean -n name.csv
!csvclean -n odonata.csv

No errors.
No errors.


Let's try it on a CSV file with syntax errors. Download bad.csv file:

In [17]:
!curl -O https://raw.githubusercontent.com/wireservice/csvkit/0.9.1/examples/bad.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    94  100    94    0     0    168      0 --:--:-- --:--:-- --:--:--   723


Take a look at it:

In [18]:
!csvlook bad.csv

|-----------+-----------------------+----------+----------------|
|  column_a | column_b              | column_c  |
|-----------+-----------------------+----------+----------------|
|  1        | 27                    |          | I'm too long!  |
|           | I'm too short!         |
|  0        | mixed types.... uh oh | 17        |
|-----------+-----------------------+----------+----------------|


See what csvclean will tell us:

In [19]:
!csvclean -n bad.csv

Line 1: Expected 3 columns, found 4 columns
Line 2: Expected 3 columns, found 2 columns


We can run in2csv for bad.csv to standardize the formatting:

In [20]:
!in2csv bad.csv | csvlook

|-----------+-----------------------+-----------|
|  column_a | column_b              | column_c  |
|-----------+-----------------------+-----------|
|  1        | 27                    |           |
|           | I'm too short!        |           |
|  0        | mixed types.... uh oh | 17        |
|-----------+-----------------------+-----------|


### csvcut
csvcut filters and truncates a CSV file.

Let's see all the columns in name.csv:

In [21]:
!csvcut -n name.csv

  1: Rank
  2: Year
  3: Name
  4: Frequency


If we only want to take a look at columns 1, 2, and 3:

In [22]:
!csvcut -c 1,2,3 name.csv | head | csvlook

|-------+------+------------|
|  Rank | Year | Name       |
|-------+------+------------|
|  1    | 1980 | Jennifer   |
|  2    | 1980 | Amanda     |
|  3    | 1980 | Melissa    |
|  4    | 1980 | Sarah      |
|  5    | 1980 | Jessica    |
|  6    | 1980 | Nicole     |
|  7    | 1980 | Elizabeth  |
|  8    | 1980 | Michelle   |
|  9    | 1980 | Amy        |
|-------+------+------------|


Do you remember how ugly the output of csvlook odonata.csv looks like? csvlook can be much more useful as a final operation when piping through other csvkit tools, especially after csvcut. Let's get all the column names in odonata.csv:

In [23]:
!csvcut -n odonata.csv

  1: Common Name
  2: Family
  3: Scientific Name
  4: State
  5: County
  6: Location
  7: Lat.
  8: Long.
  9: Elev.
 10: Date
 11: Collector
 12: #
 13: Repository
 14: Determination


Let's see the unique statistics to help us select more interesting columns:

In [24]:
!csvstat --unique odonata.csv

  1. Common Name: 16
  2. Family: 4
  3. Scientific Name: 16
  4. State: 1
  5. County: 1
  6. Location: 3
  7. Lat.: 3
  8. Long.: 3
  9. Elev.: 3
 10. Date: 1
 11. Collector: 1
 12. #: 9
 13. Repository: 1
 14. Determination: 1


Location, Lat., Long., Elev. all have 3 unique values. Let's see if they are corresponding: 

In [25]:
!csvcut -c 6,7,8,9 odonata.csv | csvlook

|----------------------------------------------------------+----------+------------+----------|
|  Location                                                | Lat.     | Long.      | Elev.    |
|----------------------------------------------------------+----------+------------+----------|
|  Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft  |
|  Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft  |
|  Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft  |
|  Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft  |
|  Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft  |
|  Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft  |
|  Baker Hot Springs (7 miles north of Brush Wellman Road) | 39.61039 | -112.73041 | 4628 ft  |
|  Fish Springs National Wildlife Refuge

Yes! So we can use Lat. and Long. to represent location and elevation. Let's select more interesting columns and make the output more readable:

In [26]:
!csvcut -c 1,2,3,7,8,12 odonata.csv | csvlook

|-------------------------+----------------+--------------------------+----------+------------+-----|
|  Common Name            | Family         | Scientific Name          | Lat.     | Long.      | #   |
|-------------------------+----------------+--------------------------+----------+------------+-----|
|  Common Green Darner    | Aeshnidae      | Anax junius              | 39.61039 | -112.73041 | 2   |
|  Paiute Dancer          | Coenagrionidae | Argia alberta            | 39.61039 | -112.73041 | 11  |
|  Black-fronted Forktail | Coenagrionidae | Ischnura denticollis     | 39.61039 | -112.73041 | 15  |
|  Western Pondhawk       | Libellulidae   | Erythemis collocata      | 39.61039 | -112.73041 | 3   |
|  Comanche Skimmer       | Libellulidae   | Libellula comanche       | 39.61039 | -112.73041 | 6   |
|  Desert Whitetail       | Libellulidae   | Plathemis subornata      | 39.61039 | -112.73041 | 1   |
|  Variegated Meadowhawk  | Libellulidae   | Sympetrum corruptum      | 39.61039 |

You may have noticed that there are several empty rows in odonata.csv. Let's strip them by adding -x argument:

In [27]:
!csvcut -x -c 1,2,3,7,8,12 odonata.csv | csvlook

|-------------------------+----------------+--------------------------+----------+------------+-----|
|  Common Name            | Family         | Scientific Name          | Lat.     | Long.      | #   |
|-------------------------+----------------+--------------------------+----------+------------+-----|
|  Common Green Darner    | Aeshnidae      | Anax junius              | 39.61039 | -112.73041 | 2   |
|  Paiute Dancer          | Coenagrionidae | Argia alberta            | 39.61039 | -112.73041 | 11  |
|  Black-fronted Forktail | Coenagrionidae | Ischnura denticollis     | 39.61039 | -112.73041 | 15  |
|  Western Pondhawk       | Libellulidae   | Erythemis collocata      | 39.61039 | -112.73041 | 3   |
|  Comanche Skimmer       | Libellulidae   | Libellula comanche       | 39.61039 | -112.73041 | 6   |
|  Desert Whitetail       | Libellulidae   | Plathemis subornata      | 39.61039 | -112.73041 | 1   |
|  Variegated Meadowhawk  | Libellulidae   | Sympetrum corruptum      | 39.61039 |

### csvgrep
csvgrep filters data to rows where certain columns match a pattern or a value.  

Let's search for the row relating to the year of 1980 in name.csv:

In [28]:
!csvgrep -c 2 -m 1980 name.csv | csvlook

|-------+------+-----------+------------|
|  Rank | Year | Name      | Frequency  |
|-------+------+-----------+------------|
|  1    | 1980 | Jennifer  | 3017       |
|  2    | 1980 | Amanda    | 1750       |
|  3    | 1980 | Melissa   | 1566       |
|  4    | 1980 | Sarah     | 1390       |
|  5    | 1980 | Jessica   | 1373       |
|  6    | 1980 | Nicole    | 1336       |
|  7    | 1980 | Elizabeth | 1221       |
|  8    | 1980 | Michelle  | 1170       |
|  9    | 1980 | Amy       | 1013       |
|  10   | 1980 | Tiffany   | 941        |
|  11   | 1980 | Angela    | 924        |
|  12   | 1980 | Kelly     | 895        |
|  13   | 1980 | Heather   | 893        |
|  14   | 1980 | Kimberly  | 854        |
|  15   | 1980 | Lisa      | 845        |
|  16   | 1980 | Stephanie | 839        |
|  17   | 1980 | Christina | 778        |
|  18   | 1980 | Rebecca   | 757        |
|  19   | 1980 | Laura     | 756        |
|  20   | 1980 | Erin      | 736        |
|  21   | 1980 | Mary      | 655  

Search for the row relating to the name starting with "B" in name.csv:

In [29]:
!csvgrep -c 3 -r "^B" name.csv | csvlook

|-------+------+----------+------------|
|  Rank | Year | Name     | Frequency  |
|-------+------+----------+------------|
|  18   | 1985 | Brittany | 714        |
|  9    | 1986 | Brittany | 952        |
|  8    | 1987 | Brittany | 1056       |
|  6    | 1988 | Brittany | 1223       |
|  4    | 1989 | Brittany | 1562       |
|  3    | 1990 | Brittany | 1533       |
|  4    | 1991 | Brittany | 1286       |
|  6    | 1992 | Brittany | 1065       |
|  7    | 1993 | Brittany | 953        |
|  9    | 1994 | Brittany | 825        |
|  16   | 1995 | Brittany | 670        |
|  20   | 1995 | Brianna  | 538        |
|  19   | 1996 | Brittany | 582        |
|  21   | 1996 | Brianna  | 527        |
|  20   | 1997 | Brianna  | 534        |
|  23   | 1997 | Brittany | 476        |
|  20   | 1998 | Brianna  | 497        |
|  20   | 1999 | Brianna  | 522        |
|  19   | 2000 | Brianna  | 533        |
|  21   | 2001 | Brianna  | 484        |
|  20   | 2002 | Brianna  | 459        |
|  18   | 2003 |

### csvsort
csvsort sorts a CSV file.  

We can sort name.csv by the "Rank" column:

In [30]:
!csvsort -c Rank name.csv | csvlook | head -n100

|-------+------+-----------+------------|
|  Rank | Year | Name      | Frequency  |
|-------+------+-----------+------------|
|  1    | 1980 | Jennifer  | 3017       |
|  1    | 1981 | Jennifer  | 2920       |
|  1    | 1982 | Jennifer  | 2809       |
|  1    | 1983 | Jennifer  | 2636       |
|  1    | 1984 | Jennifer  | 2528       |
|  1    | 1985 | Jennifer  | 2096       |
|  1    | 1986 | Jessica   | 2198       |
|  1    | 1987 | Ashley    | 2565       |
|  1    | 1988 | Jessica   | 2275       |
|  1    | 1989 | Ashley    | 2308       |
|  1    | 1990 | Jessica   | 2117       |
|  1    | 1991 | Jessica   | 2075       |
|  1    | 1992 | Jessica   | 1816       |
|  1    | 1993 | Jessica   | 1789       |
|  1    | 1994 | Jessica   | 1640       |
|  1    | 1995 | Jessica   | 1429       |
|  1    | 1996 | Jessica   | 1265       |
|  1    | 1997 | Emily     | 1213       |
|  1    | 1998 | Emily     | 1211       |
|  1    | 1999 | Emily     | 1210       |
|  1    | 2000 | Emily     | 1205 

### csvjoin
csvjoin merges CSV files on a specified column or columns.  

Let's seperate the name.csv to two different CSV files:

In [31]:
!csvcut -c 1,2,3 name.csv | csvgrep -c 2 -m 1980 > name_1.csv
!csvcut -c 1,2,4 name.csv | csvgrep -c 2 -m 1980 | csvcut -c 1,3 > name_2.csv
!csvlook name_1.csv
!csvlook name_2.csv

|-------+------+------------|
|  Rank | Year | Name       |
|-------+------+------------|
|  1    | 1980 | Jennifer   |
|  2    | 1980 | Amanda     |
|  3    | 1980 | Melissa    |
|  4    | 1980 | Sarah      |
|  5    | 1980 | Jessica    |
|  6    | 1980 | Nicole     |
|  7    | 1980 | Elizabeth  |
|  8    | 1980 | Michelle   |
|  9    | 1980 | Amy        |
|  10   | 1980 | Tiffany    |
|  11   | 1980 | Angela     |
|  12   | 1980 | Kelly      |
|  13   | 1980 | Heather    |
|  14   | 1980 | Kimberly   |
|  15   | 1980 | Lisa       |
|  16   | 1980 | Stephanie  |
|  17   | 1980 | Christina  |
|  18   | 1980 | Rebecca    |
|  19   | 1980 | Laura      |
|  20   | 1980 | Erin       |
|  21   | 1980 | Mary       |
|  22   | 1980 | Jamie      |
|  23   | 1980 | Megan      |
|  24   | 1980 | Rachel     |
|  25   | 1980 | Sara       |
|-------+------+------------|
|-------+------------|
|  Rank | Frequency  |
|-------+------------|
|  1    | 3017       |
|  2    | 1750       |
|  3    | 1566 

Now we use the default inner join to join name_1.csv and name_2.csv:

In [32]:
!csvjoin -c Rank name_1.csv name_2.csv | csvlook

|-------+------+-----------+------+------------|
|  Rank | Year | Name      | Rank | Frequency  |
|-------+------+-----------+------+------------|
|  1    | 1980 | Jennifer  | 1    | 3017       |
|  2    | 1980 | Amanda    | 2    | 1750       |
|  3    | 1980 | Melissa   | 3    | 1566       |
|  4    | 1980 | Sarah     | 4    | 1390       |
|  5    | 1980 | Jessica   | 5    | 1373       |
|  6    | 1980 | Nicole    | 6    | 1336       |
|  7    | 1980 | Elizabeth | 7    | 1221       |
|  8    | 1980 | Michelle  | 8    | 1170       |
|  9    | 1980 | Amy       | 9    | 1013       |
|  10   | 1980 | Tiffany   | 10   | 941        |
|  11   | 1980 | Angela    | 11   | 924        |
|  12   | 1980 | Kelly     | 12   | 895        |
|  13   | 1980 | Heather   | 13   | 893        |
|  14   | 1980 | Kimberly  | 14   | 854        |
|  15   | 1980 | Lisa      | 15   | 845        |
|  16   | 1980 | Stephanie | 16   | 839        |
|  17   | 1980 | Christina | 17   | 778        |
|  18   | 1980 | Reb

We need to remove the extra Rank column manually if we don't need it:

In [33]:
!csvjoin -c Rank name_1.csv name_2.csv | csvcut -c 1,2,3,5 | csvlook

|-------+------+-----------+------------|
|  Rank | Year | Name      | Frequency  |
|-------+------+-----------+------------|
|  1    | 1980 | Jennifer  | 3017       |
|  2    | 1980 | Amanda    | 1750       |
|  3    | 1980 | Melissa   | 1566       |
|  4    | 1980 | Sarah     | 1390       |
|  5    | 1980 | Jessica   | 1373       |
|  6    | 1980 | Nicole    | 1336       |
|  7    | 1980 | Elizabeth | 1221       |
|  8    | 1980 | Michelle  | 1170       |
|  9    | 1980 | Amy       | 1013       |
|  10   | 1980 | Tiffany   | 941        |
|  11   | 1980 | Angela    | 924        |
|  12   | 1980 | Kelly     | 895        |
|  13   | 1980 | Heather   | 893        |
|  14   | 1980 | Kimberly  | 854        |
|  15   | 1980 | Lisa      | 845        |
|  16   | 1980 | Stephanie | 839        |
|  17   | 1980 | Christina | 778        |
|  18   | 1980 | Rebecca   | 757        |
|  19   | 1980 | Laura     | 756        |
|  20   | 1980 | Erin      | 736        |
|  21   | 1980 | Mary      | 655  

### csvstack
csvstack stacks up the rows from multiple CSV files.  

Let's extract two parts from name.csv according to the rank value:

In [34]:
!csvgrep -c 2 -m 1980 name.csv | csvgrep -c 1 -r "^\d$" > name_a.csv
!csvgrep -c 2 -m 1980 name.csv | csvgrep -c 1 -r "^1\d$" > name_b.csv
!csvlook name_a.csv
!csvlook name_b.csv

|-------+------+-----------+------------|
|  Rank | Year | Name      | Frequency  |
|-------+------+-----------+------------|
|  1    | 1980 | Jennifer  | 3017       |
|  2    | 1980 | Amanda    | 1750       |
|  3    | 1980 | Melissa   | 1566       |
|  4    | 1980 | Sarah     | 1390       |
|  5    | 1980 | Jessica   | 1373       |
|  6    | 1980 | Nicole    | 1336       |
|  7    | 1980 | Elizabeth | 1221       |
|  8    | 1980 | Michelle  | 1170       |
|  9    | 1980 | Amy       | 1013       |
|-------+------+-----------+------------|
|-------+------+-----------+------------|
|  Rank | Year | Name      | Frequency  |
|-------+------+-----------+------------|
|  10   | 1980 | Tiffany   | 941        |
|  11   | 1980 | Angela    | 924        |
|  12   | 1980 | Kelly     | 895        |
|  13   | 1980 | Heather   | 893        |
|  14   | 1980 | Kimberly  | 854        |
|  15   | 1980 | Lisa      | 845        |
|  16   | 1980 | Stephanie | 839        |
|  17   | 1980 | Christina | 778  

Now we can stack these two CSV files:

In [35]:
!csvstack name_a.csv name_b.csv | csvlook

|-------+------+-----------+------------|
|  Rank | Year | Name      | Frequency  |
|-------+------+-----------+------------|
|  1    | 1980 | Jennifer  | 3017       |
|  2    | 1980 | Amanda    | 1750       |
|  3    | 1980 | Melissa   | 1566       |
|  4    | 1980 | Sarah     | 1390       |
|  5    | 1980 | Jessica   | 1373       |
|  6    | 1980 | Nicole    | 1336       |
|  7    | 1980 | Elizabeth | 1221       |
|  8    | 1980 | Michelle  | 1170       |
|  9    | 1980 | Amy       | 1013       |
|  10   | 1980 | Tiffany   | 941        |
|  11   | 1980 | Angela    | 924        |
|  12   | 1980 | Kelly     | 895        |
|  13   | 1980 | Heather   | 893        |
|  14   | 1980 | Kimberly  | 854        |
|  15   | 1980 | Lisa      | 845        |
|  16   | 1980 | Stephanie | 839        |
|  17   | 1980 | Christina | 778        |
|  18   | 1980 | Rebecca   | 757        |
|  19   | 1980 | Laura     | 756        |
|-------+------+-----------+------------|


We can also use an optional flag -g to add a grouping column to indicate which file each row came from:

In [36]:
!csvstack -g a,b name_a.csv name_b.csv | csvlook

|--------+------+------+-----------+------------|
|  group | Rank | Year | Name      | Frequency  |
|--------+------+------+-----------+------------|
|  a     | 1    | 1980 | Jennifer  | 3017       |
|  a     | 2    | 1980 | Amanda    | 1750       |
|  a     | 3    | 1980 | Melissa   | 1566       |
|  a     | 4    | 1980 | Sarah     | 1390       |
|  a     | 5    | 1980 | Jessica   | 1373       |
|  a     | 6    | 1980 | Nicole    | 1336       |
|  a     | 7    | 1980 | Elizabeth | 1221       |
|  a     | 8    | 1980 | Michelle  | 1170       |
|  a     | 9    | 1980 | Amy       | 1013       |
|  b     | 10   | 1980 | Tiffany   | 941        |
|  b     | 11   | 1980 | Angela    | 924        |
|  b     | 12   | 1980 | Kelly     | 895        |
|  b     | 13   | 1980 | Heather   | 893        |
|  b     | 14   | 1980 | Kimberly  | 854        |
|  b     | 15   | 1980 | Lisa      | 845        |
|  b     | 16   | 1980 | Stephanie | 839        |
|  b     | 17   | 1980 | Christina | 778        |


## 3. Use csvkit elsewhere
The command line may not be enough when solving some problems. In this section, we will see how csvkit collaborates with other tools to make it more powerful.

### 3.1. sql
csvsql and sql2csv are bridges connecting the CSV file and the SQL database.
### csvsql
csvsql generates SQL statement for a CSV file or executes those statements on a database.  

Let's save the stacked rows above into a new CSV file:

In [37]:
!csvstack -g a,b name_a.csv name_b.csv > stacked.csv

We can generate a create table statement for stacked.csv:

In [38]:
!csvsql -i sqlite stacked.csv

CREATE TABLE stacked (
	"group" VARCHAR(1) NOT NULL, 
	"Rank" INTEGER NOT NULL, 
	"Year" INTEGER NOT NULL, 
	"Name" VARCHAR(9) NOT NULL, 
	"Frequency" INTEGER NOT NULL
);


We can create a table and import data from stacked.csv:

In [39]:
!csvsql --db sqlite:///dummy.db --insert stacked.csv

(sqlite3.OperationalError) table stacked already exists [SQL: u'\nCREATE TABLE stacked (\n\t"group" VARCHAR(1) NOT NULL, \n\t"Rank" INTEGER NOT NULL, \n\t"Year" INTEGER NOT NULL, \n\t"Name" VARCHAR(9) NOT NULL, \n\t"Frequency" INTEGER NOT NULL\n)\n\n']


### sql2csv
sql2csv executes arbitrary commands against a SQL database and outputs the results as a CSV file.  

We can use sql2csv to check the data we just imported:

In [40]:
!sql2csv --db sqlite:///dummy.db --query "select * from stacked" | csvlook

|--------+------+------+-----------+------------|
|  group | Rank | Year | Name      | Frequency  |
|--------+------+------+-----------+------------|
|  a     | 1    | 1980 | Jennifer  | 3017       |
|  a     | 2    | 1980 | Amanda    | 1750       |
|  a     | 3    | 1980 | Melissa   | 1566       |
|  a     | 4    | 1980 | Sarah     | 1390       |
|  a     | 5    | 1980 | Jessica   | 1373       |
|  a     | 6    | 1980 | Nicole    | 1336       |
|  a     | 7    | 1980 | Elizabeth | 1221       |
|  a     | 8    | 1980 | Michelle  | 1170       |
|  a     | 9    | 1980 | Amy       | 1013       |
|  b     | 10   | 1980 | Tiffany   | 941        |
|  b     | 11   | 1980 | Angela    | 924        |
|  b     | 12   | 1980 | Kelly     | 895        |
|  b     | 13   | 1980 | Heather   | 893        |
|  b     | 14   | 1980 | Kimberly  | 854        |
|  b     | 15   | 1980 | Lisa      | 845        |
|  b     | 16   | 1980 | Stephanie | 839        |
|  b     | 17   | 1980 | Christina | 778        |


To get the first rank from the stacked table in sqlite database, we could run:

In [41]:
!sql2csv --db sqlite:///dummy.db --query "select * from stacked where rank=1;" | csvlook

|--------+------+------+----------+------------|
|  group | Rank | Year | Name     | Frequency  |
|--------+------+------+----------+------------|
|  a     | 1    | 1980 | Jennifer | 3017       |
|--------+------+------+----------+------------|


We can even entirely skip the database because csvsql will create one in memory for us:

In [42]:
!csvsql --query "select * from stacked where rank=1;" stacked.csv | csvlook

|--------+------+------+----------+------------|
|  group | Rank | Year | Name     | Frequency  |
|--------+------+------+----------+------------|
|  a     | 1    | 1980 | Jennifer | 3017       |
|--------+------+------+----------+------------|


Want something more interesting?  

Extract all the Rank 1 row from name.csv:

In [43]:
!csvgrep -c 1 -r "^1$" name.csv > rank1.csv
!csvlook rank1.csv

|-------+------+----------+------------|
|  Rank | Year | Name     | Frequency  |
|-------+------+----------+------------|
|  1    | 1980 | Jennifer | 3017       |
|  1    | 1981 | Jennifer | 2920       |
|  1    | 1982 | Jennifer | 2809       |
|  1    | 1983 | Jennifer | 2636       |
|  1    | 1984 | Jennifer | 2528       |
|  1    | 1985 | Jennifer | 2096       |
|  1    | 1986 | Jessica  | 2198       |
|  1    | 1987 | Ashley   | 2565       |
|  1    | 1988 | Jessica  | 2275       |
|  1    | 1989 | Ashley   | 2308       |
|  1    | 1990 | Jessica  | 2117       |
|  1    | 1991 | Jessica  | 2075       |
|  1    | 1992 | Jessica  | 1816       |
|  1    | 1993 | Jessica  | 1789       |
|  1    | 1994 | Jessica  | 1640       |
|  1    | 1995 | Jessica  | 1429       |
|  1    | 1996 | Jessica  | 1265       |
|  1    | 1997 | Emily    | 1213       |
|  1    | 1998 | Emily    | 1211       |
|  1    | 1999 | Emily    | 1210       |
|  1    | 2000 | Emily    | 1205       |
|  1    | 2001 |

Count the frequency of all the names in rank1.csv:

In [44]:
!csvsql --query "select Name, sum(Frequency) as s from rank1 group by Name order by s DESC;" rank1.csv

Name,s
Jessica,16604
Jennifer,16006
Emily,10379
Ashley,4873
ISABELLA,2804
SOPHIA,1847
OLIVIA,1706
EMILY,937


Let's use asciigraph to make things more interesting:

In [45]:
!csvsql --query "select Name, sum(Frequency) as s from rank1 group by Name order by s DESC;" rank1.csv | \
csvformat -D ":" | tail -n +2 | asciigraph

██████████████████████████████████████████████████████████████  16604  Jessica 
███████████████████████████████████████████████████████████     16006  Jennifer
██████████████████████████████████████                          10379  Emily   
██████████████████                                               4873  Ashley  
██████████                                                       2804  ISABELLA
██████                                                           1847  SOPHIA  
██████                                                           1706  OLIVIA  
███                                                               937  EMILY   


### 3.2. json
### csvjson
csvjson converts a CSV file into JSON or GeoJSON.  

Let's make a smaller slice from stacked.csv:

In [46]:
!csvcut -c 2,3,4 stacked.csv | csvgrep -c 1 -r "^[1-3]$" | csvlook

|-------+------+-----------|
|  Rank | Year | Name      |
|-------+------+-----------|
|  1    | 1980 | Jennifer  |
|  2    | 1980 | Amanda    |
|  3    | 1980 | Melissa   |
|-------+------+-----------|


Convert that small slice into JSON format:

In [47]:
!csvcut -c 2,3,4 stacked.csv | csvgrep -c 1 -r "^[1-3]$" | csvjson --indent 4

[
    {
        "Rank": "1", 
        "Year": "1980", 
        "Name": "Jennifer"
    }, 
    {
        "Rank": "2", 
        "Year": "1980", 
        "Name": "Amanda"
    }, 
    {
        "Rank": "3", 
        "Year": "1980", 
        "Name": "Melissa"
    }
]

We can provide a column name as the key to the JSON and all the values in the column must be unique:

In [48]:
!csvcut -c 2,3,4 stacked.csv | csvgrep -c 1 -r "^[1-3]$" | csvjson --indent 4 --key Rank

{
    "1": {
        "Rank": "1", 
        "Year": "1980", 
        "Name": "Jennifer"
    }, 
    "2": {
        "Rank": "2", 
        "Year": "1980", 
        "Name": "Amanda"
    }, 
    "3": {
        "Rank": "3", 
        "Year": "1980", 
        "Name": "Melissa"
    }
}

### 3.3. python
### csvpy
csvpy loads a CSV file into a reader object and then drops into a Python shell.  

Invoking csvpy would launch a Python terminal: 

>$ csvpy stacked.csv  

```python
>>>reader.next()  
[u'group', u'Rank', u'Year', u'Name', u'Frequency']
```

We can also use a CSV DictReader instead:  

>$ csvpy --dict stacked.csv -v  

```python
>>>reader.next()
{u'Frequency': u'3017',
 u'Name': u'Jennifer',
 u'Rank': u'1',
 u'Year': u'1980',
 u'group': u'a'}
```

## 4. Use csvkit as a Python library

In addition to the command-line usage, we can **import csvkit** to use it as a Python library.

In [49]:
import csvkit

Let's read a CSV file using CSVKitReader() and CSVKitDictReader():

In [50]:
with open('stacked.csv', 'rb') as f:
    reader = csvkit.CSVKitReader(f)
    for i in xrange(10):
        print reader.next()
f.close()

[u'group', u'Rank', u'Year', u'Name', u'Frequency']
[u'a', u'1', u'1980', u'Jennifer', u'3017']
[u'a', u'2', u'1980', u'Amanda', u'1750']
[u'a', u'3', u'1980', u'Melissa', u'1566']
[u'a', u'4', u'1980', u'Sarah', u'1390']
[u'a', u'5', u'1980', u'Jessica', u'1373']
[u'a', u'6', u'1980', u'Nicole', u'1336']
[u'a', u'7', u'1980', u'Elizabeth', u'1221']
[u'a', u'8', u'1980', u'Michelle', u'1170']
[u'a', u'9', u'1980', u'Amy', u'1013']


In [51]:
with open('stacked.csv', 'rb') as f:
    reader = csvkit.CSVKitDictReader(f)
    print reader.fieldnames
    for i in xrange(10):
        print reader.next()
f.close()

[u'group', u'Rank', u'Year', u'Name', u'Frequency']
{u'Frequency': u'3017', u'group': u'a', u'Name': u'Jennifer', u'Rank': u'1', u'Year': u'1980'}
{u'Frequency': u'1750', u'group': u'a', u'Name': u'Amanda', u'Rank': u'2', u'Year': u'1980'}
{u'Frequency': u'1566', u'group': u'a', u'Name': u'Melissa', u'Rank': u'3', u'Year': u'1980'}
{u'Frequency': u'1390', u'group': u'a', u'Name': u'Sarah', u'Rank': u'4', u'Year': u'1980'}
{u'Frequency': u'1373', u'group': u'a', u'Name': u'Jessica', u'Rank': u'5', u'Year': u'1980'}
{u'Frequency': u'1336', u'group': u'a', u'Name': u'Nicole', u'Rank': u'6', u'Year': u'1980'}
{u'Frequency': u'1221', u'group': u'a', u'Name': u'Elizabeth', u'Rank': u'7', u'Year': u'1980'}
{u'Frequency': u'1170', u'group': u'a', u'Name': u'Michelle', u'Rank': u'8', u'Year': u'1980'}
{u'Frequency': u'1013', u'group': u'a', u'Name': u'Amy', u'Rank': u'9', u'Year': u'1980'}
{u'Frequency': u'941', u'group': u'b', u'Name': u'Tiffany', u'Rank': u'10', u'Year': u'1980'}


We can also write data into a CSV file. Like CSVKitReader() and CSVKitDictReader(), we can use CSVKitWriter() and CSVKitDictWriter(). For example, let's create an activity plan:

In [52]:
import calendar
import datetime
import random
import pandas as pd
with open('activity.csv', 'wb') as f:
    writer = csvkit.CSVKitWriter(f)
    writer.writerow(['Date', 'Day of Week', 'Activity'])
    start_date = 20161101
    act_set = ['Running', 'Swimming', 'Jogging', 'Walking']
    for i in xrange(10):
        data_date = str(start_date + i)
        data_date_dt = datetime.datetime.strptime(data_date, '%Y%m%d')
        data_day_of_week = calendar.day_name[data_date_dt.weekday()]
        data_act = act_set[random.randint(0, 2)]
        writer.writerow([data_date, data_day_of_week, data_act])
f.close()
df = pd.read_csv('activity.csv')
print df

       Date Day of Week  Activity
0  20161101     Tuesday   Running
1  20161102   Wednesday   Jogging
2  20161103    Thursday   Running
3  20161104      Friday   Running
4  20161105    Saturday   Jogging
5  20161106      Sunday   Jogging
6  20161107      Monday   Jogging
7  20161108     Tuesday  Swimming
8  20161109   Wednesday   Running
9  20161110    Thursday  Swimming


When using CSVKitDictWriter(), we need to specify the fieldnames:

In [53]:
with open('activity_dict.csv', 'wb') as f:
    fields = ['Date', 'Day of Week', 'Activity']
    writer = csvkit.CSVKitDictWriter(f, fieldnames=fields)
    writer.writeheader() # don't forget this line!
    start_date = 20161101
    act_set = ['Singing', 'Dancing', 'Reading']
    for i in xrange(10):
        dct = {}
        data_date = str(start_date + i)
        dct[fields[0]] = data_date
        data_date_dt = datetime.datetime.strptime(data_date, '%Y%m%d')
        data_day_of_week = calendar.day_name[data_date_dt.weekday()]
        dct[fields[1]] = data_day_of_week
        data_act = act_set[random.randint(0, 2)]
        dct[fields[2]] = data_act
        writer.writerow(dct)
f.close()
df = pd.read_csv('activity_dict.csv')
print df

       Date Day of Week Activity
0  20161101     Tuesday  Reading
1  20161102   Wednesday  Reading
2  20161103    Thursday  Singing
3  20161104      Friday  Dancing
4  20161105    Saturday  Reading
5  20161106      Sunday  Reading
6  20161107      Monday  Reading
7  20161108     Tuesday  Singing
8  20161109   Wednesday  Reading
9  20161110    Thursday  Dancing


## 4. Summary and references
This tutorial introduces how to use csvkit as command-line tools and as a Python library. If you want to explore more about csvkit, you could:

1. Go to [documentation](http://csvkit.readthedocs.io/en/540/index.html) for more details about the usage of csvkit.  
2. Go to [DATA.GOV](https://catalog.data.gov/dataset) for more real datasets of a variety of topics.